zilliqa/src/block_store.rs

```diff
@@ -18,18 +19,109 @@ use crate::{
+    highest_known_view: u64,
+    /// The maximum view of blocks we've sent a request for.
+    requested_view: u64,
+    /// The maximum number of blocks to send requests for at a time.
+    max_blocks_in_flight: u64,
+    /// The maximum number of blocks to request at a time.
+    batch_size: u64,
+    /// Buffered block proposals, indexed by their parent hash.
+    //buffered: HashMap<Hash, Proposal>,
+    buffered: LruCache<Hash, Proposal>,
     message_sender: MessageSender,
 }
 
 impl BlockStore {
-    pub fn new(db: Arc<Db>, message_sender: MessageSender) -> Result<Self> {
+    pub fn new(config: &NodeConfig, db: Arc<Db>, message_sender: MessageSender) -> Result<Self> {
         Ok(BlockStore {
             db,
             block_cache: RefCell::new(LruCache::new(NonZeroUsize::new(5).unwrap())),
+            highest_known_view: 0,
+            requested_view: 0,
+            max_blocks_in_flight: config.max_blocks_in_flight,
+            batch_size: config.block_request_batch_size,
+            buffered: LruCache::new(
+                NonZeroUsize::new(config.max_blocks_in_flight as usize + 100).unwrap(),
+            ),
             message_sender,
         })
     }
 
+    /// Buffer a block proposal whose parent we don't yet know about.
+    pub fn buffer_proposal(
+        &mut self,
+        proposal: Proposal,
+        random_peer: Option<PeerId>,
+    ) -> Result<()> {
+        let view = proposal.view();
+
+        // We want to be careful about buffering too many proposals. There is no guarantee or proof that any of them
+        // will eventually form our canonical chain. Therefore we limit the number of buffered proposals to
+        // `max_blocks_in_flight + 100`. This number is chosen because it guarantees we will always be able to handle
+        // pending requests made by ourselves that arrive out-of-order, while also giving some extra space for newly
+        // created blocks that arrive while we are syncing.
+        self.buffered.push(proposal.header.parent_hash, proposal);
+
+        // If this is the highest block we've seen, remember its view.
+        if view > self.highest_known_view {
+            trace!(view, "new highest known view");
+            self.highest_known_view = view;
+        }
+
+        self.request_missing_blocks(random_peer)?;
+
+        Ok(())
+    }
+
+    pub fn request_missing_blocks(&mut self, random_peer: Option<PeerId>) -> Result<()> {
+        // Get the highest view we currently have committed to our chain.
+        let current_view = self
+            .get_block_by_number(
+                self.db
+                    .get_highest_block_number()?
+                    .ok_or_else(|| anyhow!("no highest block"))?,
+            )?
+            .ok_or_else(|| anyhow!("missing highest block"))?
+            .view();
+        // We can consider all blocks we currently have committed to already be requested.
+        if self.requested_view < current_view {
```

Comment:
Really? Why? Presumably because we've already got them, so perhaps clearer to say
```
// We've already got any block we've previously committed, so don't request it again
```
or have I misunderstood?

Reply:
That's much better, thank you :)

Reply:
:-)